pkg/forwarders/tlsforwarder_test.go

```diff
@@ -75,126 +131,72 @@ f3cb9gYaLWdmvkx8p3g=
 		&tls.Config{
-			RootCAs:      roots,
-			Certificates: []tls.Certificate{tlsCert},
-			// Tried to actually verify the certificates in this test but was getting the following error
-			// event though 127.0.0.1 and localhost were added to alt_names in openssl config:
-			// failed to connect to reporting server: x509: cannot validate certificate for 127.0.0.1 because it doesn't contain any IP SANs
+			Certificates:       []tls.Certificate{tlsCert},
 			InsecureSkipVerify: true, //nolint:gosec
 		},
 	)
 	require.NoError(t, err)
-	require.NotNil(t, listener)
 	t.Logf("TLS server address %s", listener.Addr().String())
 
-	var wg sync.WaitGroup
-	go acceptLoop(t, listener, &wg,
-		[]string{
-			"<14>signal=test-signal;key=value;\n",
-			"<14>signal=test-signal-2;key=value;\n",
-		},
-	)
-
-	serializer := serializers.NewSemicolonDelimited()
-	require.NotNil(t, serializer)
-
-	log := logrusr.New(logrus.New())
-	tf, err := NewTLSForwarder(listener.Addr().String(), log,
-		func(c *tls.Config) {
-			c.RootCAs = roots
-			c.Certificates = []tls.Certificate{tlsCert}
-			c.InsecureSkipVerify = true
-		},
-	)
-	require.NoError(t, err)
-	require.NotNil(t, tf)
-
-	consumer := telemetry.NewConsumer(serializer, tf)
-
-	m, err := telemetry.NewManager(
-		"test-ping",
-		telemetry.OptManagerPeriod(time.Hour),
-		telemetry.OptManagerLogger(log),
-	)
-	require.NoError(t, err)
-
-	w := telemetry.NewWorkflow("test1")
-	p, err := provider.NewFixedValueProvider("test1-provider", types.ProviderReport{
-		"key": "value",
-	})
-	require.NoError(t, err)
-	w.AddProvider(p)
-	m.AddWorkflow(w)
-	require.NoError(t, m.AddConsumer(consumer))
-	require.NoError(t, m.Start())
+	return telemetryServer{
+		listener: listener,
+		done:     make(chan struct{}),
+	}
+}
 
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal"))
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal-2"))
-	wg.Wait()
+func (ts telemetryServer) Addr() string {
+	return ts.listener.Addr().String()
+}
 
-	require.NoError(t, listener.Close())
-	m.Stop()
+func (ts telemetryServer) Done() {
+	<-ts.done
 }
 
-func acceptLoop(t *testing.T, l net.Listener, wg *sync.WaitGroup, expectedData []string) {
+func (ts telemetryServer) RunAndAssertExpectedData(t *testing.T, expectedData []string) {
 	t.Log("server: accepting...")
-
-	// Accept just once because TLSForwarder persists the connection
-	conn, err := l.Accept()
-	if err != nil {
-		t.Logf("server: Accept() returned error: %v", err)
-		return
+	receivedData := make(chan string, len(expectedData))
+	go func() {
+		for {
+			conn, err := ts.listener.Accept()
+			if err != nil {
+				t.Logf("server: Accept() returned error: %v", err)
+				return
+			}
+			go handleConnection(t, conn, receivedData)
+		}
+	}()
+	defer ts.listener.Close()
+	for _, expected := range expectedData {
+		select {
+		case received := <-receivedData:
+			assert.Equal(t, expected, received)
+		case <-time.After(5 * time.Second):
```

Comment:
What would you say about using a `context.Context` parameter and using it here?

We could control the timeout at the call site then and control the cancellation (and share the context with `m.TriggerExecute(context.Background())` invocations )

Reply:
Good idea